File: c/isPalindrome.cpp
// https://leetcode.com/problems/palindrome-linked-list/

#include <iostream>

#include "ListNode.hpp"
using namespace std;

class Solution {
   public:
    bool isPalindrome(ListNode *head) {
        ListNode *slow = head;
        ListNode *fast = head;
        ListNode *prev, *temp;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        prev = slow;
        slow = slow->next;
        prev->next = nullptr;
        while (slow != nullptr) {
            temp = slow->next;
            slow->next = prev;
            prev = slow;
            slow = temp;
        }

        fast = head;
        slow = prev;
        while (slow != nullptr) {
            if (fast->val != slow->val) {
                return false;
            } else {
                fast = fast->next;
                slow = slow->next;
            }
        }

        return true;
    }
};

int main() {
    Solution solution;
    ListNode *head = new ListNode(
        1, new ListNode(
               4, new ListNode(5, new ListNode(4, new ListNode(1, nullptr)))));
    printList(head);
    bool result = solution.isPalindrome(head);
    cout << result;
    return 0;
}

File: rust/src/bin/is_palindrome.rs
// https://leetcode.com/problems/reverse-linked-list/
// 2024/03/22

mod list_node;
use crate::list_node::*;

impl Solution {
    pub fn collect(node: Option<Box<ListNode>>) -> Vec<i32> {
        let mut result: Vec<i32> = Vec::new();
        let mut node: Option<Box<ListNode>> = node;

        loop {
            match node {
                None => break,
                Some(value) => {
                    result.push(value.val);
                    node = value.next;
                }
            };
        }

        return result;
    }

    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool {
        let nodes: Vec<i32> = Solution::collect(head);
        for i in 0..nodes.len() {
            if nodes[i] != nodes[nodes.len() - i - 1] {
                return false;
            }
        }

        return true;
    }
}

pub struct Solution {}

fn main() {
    println!(
        "{:?}",
        Solution::is_palindrome(ListNode::from_array(&[1, 2, 2, 1]))
    ); // true

    println!(
        "{:?}",
        Solution::is_palindrome(ListNode::from_array(&[1, 2,]))
    ); // false
}
